include/triton/Conversion/TritonGPUToLLVM/Utility.h

```diff
@@ -463,7 +450,8 @@ inline Value getSharedMemoryBase(Location loc, RewriterBase &rewriter,
                       .getValue()
                       .getZExtValue();
   Value offVal = i32_val(offset);
-  Value base = gep(ptrTy, i8_ty, LLVM::getStackPointer(rewriter, func), offVal);
+  Value base =
```

Comment:
this is not something new, how is that related to tl.gather? Does your backend represent shared memory differently?

Reply:
Maybe there's something not out-of-sync in intel's backend. They still use `numArgs - 1` to store the smem scratch buffer pointer, while the current triton function's calling convention is:

```
raw_args.., smem scratch buffer pointer, gmem scratch buffer pointer
```

Intel's implementation
```
Value TargetInfo::getStackPointer(RewriterBase &rewriter,
                                  FunctionOpInterface funcOp) const {
  auto mod = funcOp->getParentOfType<ModuleOp>();
  LLVM::LLVMPointerType ptrTy = ptr_ty(
      rewriter.getContext(), TritonGEN::TritonGENMemorySpace::kWorkgroup);
  if (mod->getAttrOfType<IntegerAttr>("ttg.shared").getInt() == 0)
    return rewriter.create<LLVM::PoisonOp>(funcOp.getLoc(), ptrTy);
  return funcOp.getArgument(funcOp.getNumArguments() - 1);
}
````

Reply:
Yes. We want to have different calling convention for passing scratch spaces. This should be abstract concept for different target.

The major difference to the NV/AMD is that: The Intel uses the same convention for kernel entry function and non-kernel entry function for passing the scratch spaces in SLM.

We'd like have a decoupled code in common code to get the pointer of the scratch spaces of the GPU functions.

Reply:
how can the kernel take a shared memory pointer arguments? Those don't have meaning on the host. 
Is it because of a different representation for openCL kernel? Can this be resolved at the LLVM level instead. It's not great to have to duplicate code and logic between NV and AMD

Reply:
> w is that related to tl.gather?
Current Gather op lowering requires base pointer to the scratch memory. https://github.com/triton-lang/triton/blob/main/lib/Conversion/TritonGPUToLLVM/GatherOpToLLVM.cpp#L87

Reply:
It is the Intel SYCL calling convention and which is inherited from the OpenCL. 

There are two ways to allocate the SLM for kernel to run on GPU. We uses the dynamic allocation for the front-end compiler (Triton) because the backend (IGC, Intel Graphic Compiler) may use the static allocation for some micro-kernel-function. We don't want to have the conflicting of the SLM allocation which is easy to debugging issue.

Here is the launcher of the Intel to allocate the scratch space in SLM on host side.
https://github.com/intel/intel-xpu-backend-for-triton/blob/6e466f419a9981fc4b8e2649f050d9548afe8675/third_party/intel/backend/driver.py#L352

The basic idea is that we'd like decouple the calling convention for passing the scratch space. It is more natural that scratch space is a concept on different memory, Private Memory, Shared Local Memory or Global Memory.

Reply:
> The basic idea is that we'd like to decouple the calling convention for passing the scratch space. It is more natural that scratch space is a concept on different memory, Private Memory, Shared Local Memory or Global Memory.

If this is the case, more changes have to be made. First, a method to retrieve the global scratch memory has to exist in intel's backend, at least an interface even if not implemented. Second, it has to be made clear that the calling convention is not fixed in comments. Third, I'm not sure if implementing all the stuff in TargetInfo is a good idea since it duplicates. Maybe it has to be refactored using either a helper class or CRTP

Reply:
Make sense. We'd like to revisit the changes inside Intel and create a new one.